src/v/storage/disk_log_impl.cc

```diff
@@ -557,43 +557,71 @@ model::term_id disk_log_impl::term() const {
     }
     return _segs.back()->offsets().term;
 }
+static offset_stats empty_offsets(model::offset start_offset) {
+    offset_stats ret;
+    ret.start_offset = start_offset;
+    if (ret.start_offset > model::offset(0)) {
+        ret.dirty_offset = ret.start_offset - model::offset(1);
+        ret.committed_offset = ret.dirty_offset;
+    }
+    return ret;
+}
 
 offset_stats disk_log_impl::offsets() const {
```

Comment:
I tried to rework the function, but I still don't understand offsets, I guess:
```c++
offset_stats disk_log_impl::offsets() const {
    segment_set::type prev_seg;
    segment_set::type term_start;
    segment_set::type last_committed;
    /**
     * We have to iterate backward over segments set as we have to find last
     * non empty segment, last segment with initialized committed offset and
     * first offset of current term
     */
    const auto rbegin = std::make_reverse_iterator(_segs.cbegin());
    const auto rend = std::find_if(
      std::make_reverse_iterator(_segs.cend()), rbegin, [](const auto& seg) {
          return !seg->empty();
      });

    for (auto it = rend; it != rbegin && !(last_committed && term_start);
         prev_seg = *it++) {
        const auto seg = *it;
        const auto& segment_offsets = seg->offsets();

        // last_committed is the latest segment with a non-negative committed
        // offset
        if (
          !last_committed
          && segment_offsets.committed_offset >= model::offset(0)) {
            last_committed = seg;
        }

        // term_start is the earliest segment with current term
        if (
          !term_start && prev_seg
          && segment_offsets.term < (*rend)->offsets().term) {
            term_start = prev_seg;
        }
    }
    // Set the term to prev since it may not have rolled over
    term_start = term_start ? term_start : prev_seg;
    // I don't understand why the tests fail without this line
    last_committed = last_committed ? last_committed : prev_seg;

    if (rend == rbegin) {
        return empty_offsets(_start_offset);
    }

    // I also don't understand why this assert fails.
    vassert(
      term_start->offsets().term == (*rend)->offsets().term
        && last_committed->offsets().committed_offset() >= 0,
      "");

    // we have valid begin and end
    const auto& bof = _segs.front()->offsets();
    const auto& eof = (*rend)->offsets();
    // term start
    const auto term_start_offset = term_start->offsets().base_offset;

    const auto start_offset = _start_offset() >= 0 ? _start_offset
                                                   : bof.base_offset;

    return storage::offset_stats{
      .start_offset = start_offset,

      .committed_offset = last_committed->offsets().committed_offset,
      .committed_offset_term = last_committed->offsets().term,

      .dirty_offset = eof.dirty_offset,
      .dirty_offset_term = eof.term,
      .last_term_start_offset = term_start_offset,
    };
}
```

Reply:
I guess it's possible for nothing to have been committed?